subgraphs/rocket-pool/src/checkpoints/nodeUtilities.ts

```diff
@@ -0,0 +1,382 @@
+  }
+
+  /**
+   * Calculates and returns the maximum RPL a node operator needs as collateral for a minipool.
+   */
+  public getMaximumRPLForNewMinipool(
+    nodeDepositAmount: BigInt,
+    maximumETHCollateralRatio: BigInt,
+    rplPrice: BigInt
+  ): BigInt {
+    if (
+      nodeDepositAmount == BigInt.fromI32(0) ||
+      maximumETHCollateralRatio == BigInt.fromI32(0) ||
+      rplPrice == BigInt.fromI32(0)
+    )
+      return BigInt.fromI32(0);
+    return nodeDepositAmount.times(maximumETHCollateralRatio).div(rplPrice);
+  }
+
+  /**
+   * Updates the network checkpoint based on the state of the node.
+   * E.g. Increment running total (effective) RPL staked for this network checkpoint, ...
+   */
+  public updateNetworkNodeBalanceCheckpointForNode(
+    networkCheckpoint: NetworkNodeBalanceCheckpoint,
+    node: Node
+  ): void {
+    // Update total number of nodes registered.
+    networkCheckpoint.nodesRegistered = networkCheckpoint.nodesRegistered.plus(
+      BigInt.fromI32(1)
+    );
+
+    // Update total number of oracle nodes registered if needed.
+    if (node.isOracleNode) {
+      networkCheckpoint.oracleNodesRegistered =
+        networkCheckpoint.oracleNodesRegistered.plus(BigInt.fromI32(1));
+    }
+
+    // Update total (effective) RPL staked.
+    networkCheckpoint.rplStaked = networkCheckpoint.rplStaked.plus(
+      node.rplStaked
+    );
+    networkCheckpoint.effectiveRPLStaked =
+      networkCheckpoint.effectiveRPLStaked.plus(node.effectiveRPLStaked);
+
+    // Update the total ETH RPL Slashed up to this checkpoint.
+    networkCheckpoint.totalRPLSlashed = networkCheckpoint.totalRPLSlashed.plus(
+      node.totalRPLSlashed
+    );
+
+    // Update total RPL rewards claimed up to this checkpoint.
+    networkCheckpoint.totalODAORewardsClaimed =
+      networkCheckpoint.totalODAORewardsClaimed.plus(
+        node.totalODAORewardsClaimed
+      );
+    networkCheckpoint.totalNodeRewardsClaimed =
+      networkCheckpoint.totalNodeRewardsClaimed.plus(
+        node.totalNodeRewardsClaimed
+      );
+
+    // Update total number of minipools per state.
+    networkCheckpoint.queuedMinipools = networkCheckpoint.queuedMinipools.plus(
+      node.queuedMinipools
+    );
+    networkCheckpoint.stakingMinipools =
+      networkCheckpoint.stakingMinipools.plus(node.stakingMinipools);
+    networkCheckpoint.stakingUnbondedMinipools =
+      networkCheckpoint.stakingUnbondedMinipools.plus(
+        node.stakingUnbondedMinipools
+      );
+    networkCheckpoint.withdrawableMinipools =
+      networkCheckpoint.withdrawableMinipools.plus(node.withdrawableMinipools);
+    networkCheckpoint.totalFinalizedMinipools =
+      networkCheckpoint.totalFinalizedMinipools.plus(
+        node.totalFinalizedMinipools
+      );
+
+    updateProtocolAndPoolRewardsTvl(
+      networkCheckpoint.block,
+      networkCheckpoint.blockTime,
+      networkCheckpoint.effectiveRPLStaked
+    );
```

Comment:
A lot of the functions in this class don't save anything?

Reply:
This code is from the original Data Nexus subgraph. They don't use .save(), but this function is one of the core drivers of the subgraph so I have to assume they have it working in some way without calling .save() on the nodecheckpoint.